File: scripts/parse-local-html.js
/**
 * URL Extractor for Local HTML File
 * 
 * This script extracts URLs from the local copy of the AI for Education website
 * and formats them for inclusion in our studies.csv file.
 */

const fs = require('fs');
const path = require('path');
const cheerio = require('cheerio');
const csv = require('csv-parser');
const createCsvWriter = require('csv-writer').createObjectCsvWriter;

// Path to the local HTML file
const HTML_FILE_PATH = path.join('data', 'Studies Related to Generative AI in Education — AI for Education.html');

/**
 * Calculate string similarity between two strings
 * @param {string} a - First string
 * @param {string} b - Second string
 * @returns {number} - Similarity score between 0 and 1
 */
function calculateSimilarity(a, b) {
  const aLower = a.toLowerCase();
  const bLower = b.toLowerCase();
  
  // Simple includes check
  if (aLower.includes(bLower) || bLower.includes(aLower)) {
    return 0.8;
  }
  
  // Count matching words
  const aWords = aLower.split(/\s+/);
  const bWords = bLower.split(/\s+/);
  const matchingWords = aWords.filter(word => bWords.includes(word)).length;
  
  if (matchingWords > 0) {
    const matchRatio = matchingWords / Math.max(aWords.length, bWords.length);
    return 0.5 + (matchRatio * 0.5); // Scale to 0.5-1.0 range if there are matching words
  }
  
  return 0.0;
}

/**
 * Extract study links from the local HTML file
 * @returns {Promise<Array>} - Array of study title and URL objects
 */
async function extractStudyLinks() {
  console.log(`Reading local HTML file: ${HTML_FILE_PATH}`);
  
  try {
    // Check if file exists
    if (!fs.existsSync(HTML_FILE_PATH)) {
      throw new Error(`The file ${HTML_FILE_PATH} does not exist`);
    }
    
    // Read the HTML file
    const htmlContent = fs.readFileSync(HTML_FILE_PATH, 'utf-8');
    const $ = cheerio.load(htmlContent);
    const studyLinks = [];
    
    // First pass: Find all h4 tags which seem to be section headers
    $('h4').each((i, header) => {
      const category = $(header).text().trim();
      console.log(`\nProcessing category: ${category}`);
      
      // Find all paragraphs and elements after this header but before the next h4
      $(header).nextUntil('h4').each((j, element) => {
        // Look for study titles (usually in bold/strong tags)
        $(element).find('strong').each((k, titleElem) => {
          const title = $(titleElem).text().trim();
          
          // Skip empty titles or category headers
          if (!title || title.endsWith(':') || title === category || title.length < 5) {
            return;
          }
          
          // Try to find associated link - look at parent paragraph
          const parent = $(titleElem).parent();
          const links = parent.find('a');
          
          if (links.length > 0) {
            const url = $(links[0]).attr('href');
            console.log(`Found: ${title} -> ${url}`);
            
            studyLinks.push({
              title,
              url
            });
          }
        });
      });
    });
    
    // Second pass: Look for direct paragraphs with bold text and links
    $('p').each((i, paragraph) => {
      const strongTags = $(paragraph).find('strong');
      
      // Skip paragraphs without bold text
      if (strongTags.length === 0) {
        return;
      }
      
      strongTags.each((j, strong) => {
        const title = $(strong).text().trim();
        
        // Skip empty titles or those that seem to be headers
        if (!title || title.endsWith(':') || title.length < 5) {
          return;
        }
        
        // Check if title already exists in our list
        const exists = studyLinks.some(link => link.title === title);
        if (exists) {
          return;
        }
        
        // Look for links in this paragraph
        const links = $(paragraph).find('a');
        if (links.length > 0) {
          const url = $(links[0]).attr('href');
          console.log(`Found additional: ${title} -> ${url}`);
          
          studyLinks.push({
            title,
            url
          });
        }
      });
    });
    
    console.log(`\nFound ${studyLinks.length} study links`);
    return studyLinks;
  } catch (error) {
    console.error('Error extracting study links:', error);
    return [];
  }
}

/**
 * Update our studies CSV with the extracted URLs
 * @param {Array} studyLinks - Array of study title and URL objects
 */
async function updateStudiesCsv(studyLinks) {
  const INPUT_CSV = path.join('data', 'studies.csv');
  const OUTPUT_CSV = path.join('data', 'studies-with-links.csv');
  
  const studies = [];
  
  // Read current studies
  fs.createReadStream(INPUT_CSV)
    .pipe(csv())
    .on('data', (data) => studies.push(data))
    .on('end', () => {
      // Match studies with extracted links based on title similarity
      studies.forEach(study => {
        let bestMatch = null;
        let highestSimilarity = 0.6; // Lower threshold to catch more matches
        
        const studyTitle = study.title;
        console.log(`\nLooking for matches for: ${studyTitle}`);
        
        studyLinks.forEach(link => {
          const similarity = calculateSimilarity(studyTitle, link.title);
          console.log(`  Comparing with '${link.title}' - similarity: ${similarity.toFixed(2)}`);
          
          if (similarity > highestSimilarity) {
            highestSimilarity = similarity;
            bestMatch = link;
          }
        });
        
        if (bestMatch) {
          study.url = bestMatch.url;
          console.log(`✓ Best match: '${bestMatch.title}' -> ${bestMatch.url}`);
        } else {
          console.log(`✗ No good match found`);
        }
      });
      
      // Write updated CSV
      const csvWriter = createCsvWriter({
        path: OUTPUT_CSV,
        header: Object.keys(studies[0]).map(id => ({ id, title: id }))
      });
      
      csvWriter.writeRecords(studies)
        .then(() => {
          console.log(`\nCSV file was written successfully to ${OUTPUT_CSV}`);
        });
    });
}

/**
 * Main function to run the extractor
 */
async function main() {
  console.log('Starting URL extractor for local HTML file...');
  
  const studyLinks = await extractStudyLinks();
  
  if (studyLinks.length > 0) {
    await updateStudiesCsv(studyLinks);
  } else {
    console.log('No links found to update CSV');
  }
}

main(); 

File: scripts/parse_local_html.py
#!/usr/bin/env python3
"""
URL Extractor for Local HTML File

This script extracts URLs from the local copy of the AI for Education website
and formats them for inclusion in our studies.csv file.
"""

from bs4 import BeautifulSoup
import csv
import os
from difflib import SequenceMatcher

# Path to the local HTML file
HTML_FILE_PATH = 'data/Studies Related to Generative AI in Education — AI for Education.html'

def calculate_similarity(a, b):
    """Calculate string similarity between two strings."""
    return SequenceMatcher(None, a.lower(), b.lower()).ratio()

def extract_study_links():
    """Extract study links from the local HTML file."""
    print(f"Reading local HTML file: {HTML_FILE_PATH}")
    try:
        # Check if file exists
        if not os.path.exists(HTML_FILE_PATH):
            raise FileNotFoundError(f"The file {HTML_FILE_PATH} does not exist")
        
        # Read the HTML file
        with open(HTML_FILE_PATH, 'r', encoding='utf-8') as f:
            html_content = f.read()
        
        soup = BeautifulSoup(html_content, 'html.parser')
        study_links = []
        
        # Find all h4 tags which seem to be section headers in the document
        section_headers = soup.find_all('h4')
        
        for header in section_headers:
            # The structure appears to have h4 headers for categories followed by study entries
            category = header.get_text().strip()
            print(f"\nProcessing category: {category}")
            
            # Find study entries after this header
            current_element = header.next_sibling
            
            while current_element and (not isinstance(current_element, type(header)) or current_element.name != 'h4'):
                # Look for study titles (bold text) and links
                if hasattr(current_element, 'find_all'):
                    # Find study titles (usually in bold)
                    study_titles = current_element.find_all('strong')
                    
                    for title_elem in study_titles:
                        title = title_elem.get_text().strip()
                        
                        # Skip empty titles or category headers
                        if not title or title.endswith(':') or title == category:
                            continue
                        
                        # Try to find associated link - look at parent paragraph and sibling elements
                        parent = title_elem.parent
                        links = parent.find_all('a')
                        
                        if links:
                            url = links[0].get('href', '')
                            print(f"Found: {title} -> {url}")
                            
                            study_links.append({
                                'title': title,
                                'url': url
                            })
                
                if current_element is None:
                    break
                
                # Move to next element
                current_element = current_element.next_sibling
        
        # Also look for links and titles in a more direct way as fallback
        paragraphs = soup.find_all('p')
        for p in paragraphs:
            strong_tags = p.find_all('strong')
            
            # Skip paragraphs without bold text (likely not study titles)
            if not strong_tags:
                continue
                
            for strong in strong_tags:
                title = strong.get_text().strip()
                
                # Skip empty titles or those that seem to be headers
                if not title or title.endswith(':') or len(title) < 5:
                    continue
                
                # Check if this title is already in our list
                if any(link['title'] == title for link in study_links):
                    continue
                
                # Look for links in this paragraph
                links = p.find_all('a')
                if links:
                    url = links[0].get('href', '')
                    print(f"Found additional: {title} -> {url}")
                    
                    study_links.append({
                        'title': title,
                        'url': url
                    })
        
        print(f"\nFound {len(study_links)} study links")
        return study_links
    
    except Exception as e:
        print(f"Error extracting study links: {e}")
        return []

def update_studies_csv(study_links):
    """Update our studies CSV with the extracted URLs."""
    INPUT_CSV = 'data/studies.csv'
    OUTPUT_CSV = 'data/studies-with-links.csv'
    
    studies = []
    
    # Read current studies
    with open(INPUT_CSV, 'r', newline='', encoding='utf-8') as csvfile:
        reader = csv.DictReader(csvfile)
        field_names = reader.fieldnames
        print(f"CSV fields: {field_names}")
        for row in reader:
            studies.append(row)
    
    # Match studies with extracted links based on title similarity
    for study in studies:
        best_match = None
        highest_similarity = 0.6  # Lower threshold to catch more matches
        
        study_title = study['title']
        print(f"\nLooking for matches for: {study_title}")
        
        for link in study_links:
            similarity = calculate_similarity(study_title, link['title'])
            print(f"  Comparing with '{link['title']}' - similarity: {similarity:.2f}")
            
            if similarity > highest_similarity:
                highest_similarity = similarity
                best_match = link
        
        if best_match:
            study['url'] = best_match['url']
            print(f"✓ Best match: '{best_match['title']}' -> {best_match['url']}")
        else:
            print("✗ No good match found")
    
    # Check for and remove problematic fields
    for study in studies:
        # Remove any keys that are not in fieldnames
        problem_keys = [key for key in study.keys() if key not in field_names]
        for key in problem_keys:
            print(f"Removing problematic key: {key}")
            del study[key]
        
        # Check for None keys which can happen with malformed CSV data
        if None in study:
            print(f"Removing None key")
            del study[None]
    
    # Write updated CSV
    with open(OUTPUT_CSV, 'w', newline='', encoding='utf-8') as csvfile:
        writer = csv.DictWriter(csvfile, fieldnames=field_names)
        writer.writeheader()
        writer.writerows(studies)
    
    print(f"\nCSV file was written successfully to {OUTPUT_CSV}")

def main():
    """Main function to run the extractor."""
    print("Starting URL extractor for local HTML file...")
    study_links = extract_study_links()
    
    if study_links:
        update_studies_csv(study_links)
    else:
        print("No links found to update CSV")

if __name__ == "__main__":
    main() 